### PR TITLE
Feature/benliu/care extern

### PR DESCRIPTION
**Pull request by liu15**
Refactor to allow external template declarations

### CODE REVIEW

**Comment by liu15** [comment]
I think we want to have the ability to build with or without the external template instantiations (which will be mutually exclusive with the inlining, I think).  I wanted a review on the general methodology first, however.

In particular, the KeyValueSorter and algorithm methodologies are different (KeyValueSorter_inst.h provides a way for host codes to relatively easily create new extern template instantiations).

**Review by robinson96 — APPROVED** [review]
LGTM. Thank you for the service.

File: CMakeLists.txt [hunk 1]
```diff
@@ -31,6 +31,7 @@ set(ENABLE_IMPLICIT_CONVERSIONS ${CARE_ENABLE_IMPLICIT_CONVERSIONS} CACHE BOOL "
 option(CARE_LEGACY_COMPATIBILITY_MODE "Enable legacy compatibility mode" OFF)
 option(CARE_ENABLE_MANAGED_PTR "Enable managed_ptr aliases, tests, and reproducer" ON)
 option(CARE_DISABLE_RAJAPLUGIN "Disable use of the RAJA plugin. WILL ALSO DISABLE MEMORY MOTION." OFF)
+option(CARE_ENABLE_INLINE "Enable inlining of template functions" OFF)
```
Comment (adayton1) [inline CMakeLists.txt:34]: I think this should default to ON since I believe that will provide the behavior most people expect.

**Review by adayton1 — APPROVED** [review]
LGTM

File: src/care/KeyValueSorter.h [hunk 2]
```diff
@@ -5,16 +5,22 @@
 // SPDX-License-Identifier: BSD-3-Clause
 //////////////////////////////////////////////////////////////////////////////////////
 
+// This header includes the implementations of KeyValueSorter.
+// In very large monolithic codes, including this in too many compilation
+// units can cause linking issues (particularly device link issues) due to
+// the file size. In that case, external template instantiation should be
+// used: this file should only be included in the compilation unit containing
+// the instantiation and KeyValueSorter_decl.h (along with the extern template
+// declarations) should be included everywhere else.
```
Comment (adayton1) [inline src/care/KeyValueSorter.h:14]: Maybe mention the CMake option that controls the instantiation.

File: src/care/algorithm.h [hunk 3]
```diff
@@ -8,14 +8,18 @@
...
-#include "care/host_ptr.h"
-#include "care/local_ptr.h"
-#include "care/policies.h"
+// This header includes the implementations of the CARE algorithms.
+// In very large monolithic codes, including this in too many compilation
+// units can cause linking issues (particularly device link issues) due to
+// the file size. In that case, external template instantiation should be
+// used: this file should only be included in the compilation unit containing
+// the instantiation and algorithm_decl.h (along with the extern template
+// declarations) should be included everywhere else.
```
Comment (adayton1) [inline src/care/algorithm.h:17]: Maybe mention the CMake option that controls this.

**Review by adayton1 — APPROVED** [review]
LGTM

**Pull request merged**